File: 3_Structural_Design_Patterns/3.02_Bridge/Python_Snippet/Bridge.py
# Abstraction (Painting Tool)
class PaintingTool:
    def __init__(self, color):
        self.color = color

    def apply_color(self):
        raise NotImplementedError("Subclasses must implement apply_color")

# Concrete Implementations (Painting Tools)
class Brush(PaintingTool):
    def apply_color(self):
        print(f"Applying {self.color.get_color()} with a brush")

class Roller(PaintingTool):
    def apply_color(self):
        print(f"Applying {self.color.get_color()} with a roller")

class Sprayer(PaintingTool):
    def apply_color(self):
        print(f"Applying {self.color.get_color()} with a sprayer")

# Abstraction (Color)
class Color:
    def get_color(self):
        raise NotImplementedError("Subclasses must implement get_color")

# Concrete Implementations (Colors)
class Red(Color):
    def get_color(self):
        return "Red"

class Blue(Color):
    def get_color(self):
        return "Blue"

class Green(Color):
    def get_color(self):
        return "Green"

# Usage (Client - Painter)
def main():
    brush = Brush(Red())
    brush.apply_color()  # Output: Applying Red with a brush

    roller = Roller(Blue())
    roller.apply_color()  # Output: Applying Blue with a roller

    sprayer = Sprayer(Green())
    sprayer.apply_color()  # Output: Applying Green with a sprayer

if __name__ == "__main__":
    main()

File: 3_Structural_Design_Patterns/3.02_Bridge/Java_Snippet/Bridge.java
interface Color {
    String getColor();
}

interface PaintingTool {
    void applyColor(Color color);
}

class Brush implements PaintingTool {

    @Override
    public void applyColor(Color color) {
        System.out.println("Applying " + color + " with a brush");
    }

}

class Roller implements PaintingTool {

    @Override
    public void applyColor(Color color) {
        System.out.println("Applying " + color + " with a roller");
    }

}

class Sprayer implements PaintingTool {

    @Override
    public void applyColor(Color color) {
        System.out.println("Applying " + color + " with a sprayer");
    }

}

class Red implements Color {

    @Override
    public String getColor() {
        return "Red";
    }

}

class Green implements Color {

    @Override
    public String getColor() {
        return "Green";
    }

}

class Blue implements Color {

    @Override
    public String getColor() {
        return "Blue";
    }

}

public class Bridge {
    public static void main(String[] args) {
        PaintingTool brush = new Brush();
        Color red = new Red();
        brush.applyColor(red);

        PaintingTool sprayTool = new Sprayer();
        Color blue = new Blue();
        sprayTool.applyColor(blue);
    }
}